app/scripts/directives/keyValueEditor.js

```diff
@@ -0,0 +1,422 @@
+(function() {
+  'use strict';
+
+  angular
+    .module('openshiftConsole')
+    .directive('keyValueEditor', [
+      '$timeout',
+      '$window',
+      '$filter',
+      'keyValueEditorConfig',
+      'keyValueEditorUtils',
+      function($timeout, $window, $filter, config, utils) {
+
+        var first = utils.first;
+        var contains = utils.contains;
+        var each = utils.each;
+        var counter = 1000;
+        var timeoutDelay = 25;
+
+        var newEntry = function() {
+          return {name: '', value: ''};
+        };
+
+        var addEntry = function(entries, entry) {
+          entries && entries.push(entry || newEntry());
+        };
+
+        var addEntryWithSelectors = function(entries) {
+          entries && entries.push({
+            name: '',
+            valueFrom: {
+              selected: null,
+              key: null
+            },
+            valueFromIsEditable: true
+          });
+        };
+
+        var setFocusOn = function(selector, value) {
+          // $timeout just delays enough to ensure event/$digest resolution
+          $timeout(function() {
+            var element = first($window.document.querySelectorAll(selector));
+            if(element) {
+              element.focus();
+              // if setting value, this will set the cursor at the end of the text in the value
+              if(value) {
+                element.value = '';
+                element.value = value;
+              }
+            }
+          }, timeoutDelay);
+        };
+
+        var uniqueForKey = function(unique, $index) {
+          return 'key-value-editor-key-' + unique + '-' + $index;
+        };
+
+        var uniqueForValue = function(unique, $index) {
+          return 'key-value-editor-value-' + unique + '-' + $index;
+        };
+
+        var altTextForValueFrom = function(entry) {
+          if(entry.value) {
+            return;
+          }
+          if(!entry.valueFrom) {
+            return;
+          }
+          entry.valueIcon = 'pficon pficon-help';
+          entry.valueIconTooltip = 'This is a referenced value that will be generated when a container is created.  On running pods you can check the resolved values by going to the Terminal tab and echoing the environment variable.';
+          _.each(entry.valueFrom, function(from, refType) {
+            switch(refType) {
+              case 'configMapKeyRef':
+                var configMapObj = {
+                  kind: "ConfigMap",
+                  metadata: {
+                    name: from.name
+                  }
+                };
+                console.log('configMapObj', configMapObj);
+                entry.valueAlt = 'Set to the key ' + from.key + ' in config map <a ng-if="\'configMaps\' | canI : \'get\'" ng-href="{{configMapObj | navigateResourceURL}}">' + from.name + '</a>';
```

Comment:
@benjaminapetersen or @jwforres, this doesn't create a valid href, and I'm not sure why.  Help?